### PR TITLE
Add access to pdf components and add Mix3/MixN combinators

### PR DESCRIPTION
**Pull request by chrisburr**
- Adds a `Mix3(f1, f2, A, B, C)` combinator -> equivalent to `Mix2(f2, Mix2(f1, A, B), C)`
- Adds a `MixN(fractions, Xs)` combinator -> equivalent to chaining `Mix2` as is done in `Mix3`.
- Adds access to the pdf components (#40) like follows:

```python
with Model() as model:
    X1 = Normal(mu, sigma)
    X2 = Normal(mu, sigma)
    X = Mix2(f, X1, X2)

plt.plot(xs, model[X1].pdf(xs))
plt.plot(xs, model[X3].pdf(xs))
plt.plot(xs, model[X].pdf(xs))
```

### CODE REVIEW

File: tensorprob/distribution.py [hunk 1]
```diff
@@ -37,6 +38,30 @@ def _parse_bounds_1D(lower, upper, bounds):
...
+    distributions should set:
+        Distribution.logp: tensorflow.Tensor
+            The log probability.
+        Distribution.integral: function
+            A function with arguments (lower, upper) that returns the integral
+            of the function between the specifed bounds.
+        Distribution.depends: (optional) list of tensorflow.Tensor
+            The sub-distributions of this distribution
+
+    Most distributions are bounded automatically however some distributions,
```
Comment (ibab) [inline tensorprob/distribution.py]: automatically however → automatically, however

File: tensorprob/distribution.py [hunk 2]
```diff
@@ -37,6 +38,30 @@ def _parse_bounds_1D(lower, upper, bounds):
...
+    such as combiantors, require access to the bounds of the distribution.
```
Comment (ibab) [inline tensorprob/distribution.py]: combiantors → combinators

File: tensorprob/utilities.py [hunk 3]
```diff
@@ -1,13 +1,23 @@
-from collections import defaultdict, Iterable
+from collections import defaultdict,  Iterable, namedtuple
```
Comment (ibab) [inline tensorprob/utilities.py]: The space before `Iterable` can be removed

File: tensorprob/utilities.py [hunk 4]
```diff
@@ -59,3 +69,63 @@ def pairwise(iterable):
...
+            is_inside_bounds,
+            logp,
+            tf.fill(tf.shape(X), config.dtype(-np.inf))
+        )
+
+    return logp
+
+
+def find_common_bounds(bounds_1, bounds_2):
+    """Find a new set of boundries combining `bounds_1` and `bounds_2`
```
Comment (ibab) [inline tensorprob/utilities.py]: boundries → boundaries

File: tensorprob/distributions/combinators.py [hunk 5]
```diff
@@ -1,64 +1,112 @@
...
+    # Eqivilent to calling Mix2(Mix2(Mix2(Mix2(A, B), C), D), E)
+    assert len(fs)+1 == len(Xs)
+    fractions = [fs[0], 1-fs[0]]
+    for f in fs[1:]:
+        fractions = [f*frac for frac in fractions]
+        fractions.append(1-f)
+    return _MixN(Xs, fractions, name)
+
+
+def _MixN(Xs, fractions, name=None):
```
Comment (ibab) [inline tensorprob/distributions/combinators.py]: It's strange that `fractions` and `Xs` are reversed here

Comment (chrisburr) [inline tensorprob/distributions/combinators.py]: It is :laughing:

**Comment by ibab** [comment]
LGTM :+1:

**Pull request merged**